### ISSUE TITLE
Tracking Issue for BStr/BString

### ISSUE THREAD

**Issue by joshtriplett** [rust-lang/rust#134915]
Feature gate: `#![feature(bstr)]`

This is a tracking issue for the BStr/BString types, which represent human-readable strings that are usually, but not always, UTF-8. Unlike `&str`/`String`, these types permit non-UTF-8 contents, making them suitable for user input, non-native filenames (as Path only supports native filenames), and other applications that need to round-trip whatever data the user provides.

This was approved in ACP https://github.com/rust-lang/libs-team/issues/502 .

### Public API

```rust
// In core::bstr
#[repr(transparent)]
pub struct BStr(pub [u8]);

impl Debug for BStr { ... }
impl Display for BStr { ... }
impl Deref for BStr { type Target = [u8]; ... }
impl DerefMut for BStr { ... }
// Other trait impls from bstr, including From impls

// In alloc::bstr
#[repr(transparent)]
pub struct BString(pub Vec<u8>);

impl Debug for BString { ... }
impl Display for BString { ... }
impl Deref for BString { type Target = Vec<u8>; ... }
impl DerefMut for BString { ... }
// Other trait impls from bstr, including From impls
```

### Steps / History

- [ ] Implementation: #...
- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

### Unresolved Questions

- Should we call this `BStr`/`BString`, or `ByteStr`/`ByteString`? The former will be more familiar to users of the `bstr` crate in the ecosystem. The latter is more explicit, and avoids potential naming conflicts (making it easier to, for instance, add it to the prelude).
- Should the `Display` impl use the Unicode replacement character, or do escaping like the `Debug` impl?

**Comment by joshtriplett** [comment]
In the course of implementing this, I'm addressing https://github.com/BurntSushi/bstr/issues/190 : both the `ByteStr` and `ByteString` types will implement `Index` and `IndexMut`.

**Comment by clarfonthey** [comment]
This is mostly just a nit, but it feels appropriate that the `Debug` impl would emit a `b"..."` string rather than a `"..."` string, to more closely resemble Rust syntax. It still wouldn't include the `BStr(...)` wrapper but it makes at least a little bit of sense and shouldn't cause too many issues.

Thinking about it, we should probably do the same for `CStr` as well, prefixing the string with `c`.

I'd be fine making a PR if this seems reasonable, or not if it just feels too pedantic.

**Comment by thaliaarchi** [comment]
> **Unresolved Questions**
> - Should the `Display` impl use the Unicode replacement character, or do escaping like the `Debug` impl?

`OsStr::display()` and `Path::display()` use the Unicode replacement character. I think consistency here would be good.

**Comment by HKalbasi** [comment]
What about having a `struct ByteChar(u8)` as well? With that, we can change `b"str"` and `b'c'` types to ByteStr and ByteChar over an edition.

**Comment by jhpratt** [comment]
@HKalbasi see #110998, which is the closest you'll get realistically.

**Comment by glandium** [comment]
I don't think this type is very useful without adding methods to make it closer in API to `str` than `[u8]`. I would go as far as saying it probably shouldn't deref to `[u8]`, for two reasons:
- Methods of `[u8]` will return `&[u8]`, not ByteStr. As someone using the bstr crate, this is the single most annoying aspect of its API: that for everything you do with it, you don't get a `BStr` out.
- `[u8]` has methods with the same name as `str` methods, but with a different API. (example: split).

Well, I guess the deref makes it convenient to use where something takes a `&[u8]`, but I would argue that `str` not having a deref to `[u8]` would be an indicator that maybe ByteStr shouldn't.

**Comment by thaliaarchi** [comment]
@glandium For your first point, both of these type slice to `ByteStr` (and if https://github.com/rust-lang/rust/pull/138381 merges, will implement `SliceIndex`, so will have unchecked variants).

> [joshtriplett](https://github.com/rust-lang/rust/issues/134915#issuecomment-2568677859): In the course of implementing this, I'm addressing https://github.com/BurntSushi/bstr/issues/190 : both the `ByteStr` and `ByteString` types will implement `Index` and `IndexMut`.

**Comment by joshtriplett** [comment]
Mike Hommey wrote:
> I don't think this type is very useful without adding methods to make 
> it closer in API to `str` than `[u8]`.

I very much want to add many such methods, taking a cue from the bstr crate. The initial addition of the type aimed for minimalism, but that doesn't mean that's the final state or goal.

**Comment by jplatte** [comment]
> I very much want to add many such methods, taking a cue from the bstr crate. The initial addition of the type aimed for minimalism, but that doesn't mean that's the final state or goal.

I think the problem is that with the `Deref` impl, no extra methods shadowing slice methods can be added after stabilization.

**Comment by glandium** [comment]
> I think the problem is that with the Deref impl, no extra methods shadowing slice methods can be added after stabilization.

This is exactly what I had in mind but for some reason didn't spell out, and this being a tracking issue with no mention of anything else than the current implementation, it's hard to tell what the end goal is. Thankfully, @joshtriplett clarified, but updating the top comment would be welcome.

> I very much want to add many such methods, taking a cue from the bstr crate.

I think it would be better to take cue from `str` rather than the bstr crate. One of my biggest gripes with the bstr crate is precisely that it doesn't match the `str` API. So much that I wrote a similar crate that does (I never found the time to polish it to publish it, though, maybe I should find that time somehow).

**Comment by kornelski** [comment]
Reposting for visibility https://github.com/rust-lang/libs-team/issues/502#issuecomment-2825517369

Since `ByteStr` makes no assumptions about encoding, it can be useful for processing non-text binary data too. I suggest using byte-oriented Debug/Display instead of Unicode escapes or lossy replacement chars.

**Comment by joshtriplett** [comment]
@kornelski BStr is intended for usually-Unicode text, not for arbitrary binary data. You certainly *can* abuse it for something else, but that's not its intended purpose, so I don't think that usage motivates any changes to its implementation.

If you're looking for additional methods to work with binary data, I'd suggest proposing those for `Vec<u8>` or similar.

**Comment by kornelski** [comment]
`find(pattern)` has been proposed for `[u8]` on the internals. However, ByteStr already does exactly this in the implementation, only high-level intention is different. I'm not sure if it makes sense to have exactly the same methods with different intentions behind them:

https://internals.rust-lang.org/t/idea-add-some-binary-data-methods/22799

**Comment by futile** [comment]
> `find(pattern)` has been proposed for `[u8]` on the internals. However, ByteStr already does exactly this in the implementation, only high-level intention is different. I'm not sure if it makes sense to have exactly the same methods with different intentions behind them:

Maybe it could make sense to implement that on `[u8]` instead of `ByteStr` then, and use the Deref from ByteStr -> [u8] (just assuming it exists here, not sure) for ByteStr's api?

**Comment by clarfonthey** [comment]
`Deref` is included; it's mentioned in the issue description. The goal is to offer as few methods as possible on `ByteStr`/`ByteString`, since they mostly exist to provide alternative `Display` and `Debug` implementations.

**Comment by tmandry** [comment]
Moving my comment from https://github.com/rust-lang/libs-team/issues/550#issuecomment-2896065302 to here, where it's more appropriate.

> ByteStr was discussed in the meeting, its purpose is to explicitly opt into treating maybe-utf8 as utf8-or-replacement output, so `ByteStr::new(value)` similar to calling `value.display()`.

I'm not sure how much I buy this line of reasoning. The name `ByteStr` does not make it obvious the string is expected to be UTF8. We have several other kinds of `Str` types that are not UTF8, and the term "bytes" is often used to refer to binary data. While calling `ByteStr::new()` explicitly might cause some people to think twice about the invariants of the type, there are other ways to get one, like deserializing a struct that contains one.

This makes me think of a possible story where Rust didn't have a user's back:

* Add `String` to a struct and deserialize the struct.
* Encounter UTF8 errors in production and change the type to `ByteString`.
* Later, someone else sees the field but needs a `String` and calls `.to_string()`, inadvertently roundtripping the value through a lossy path.

Again, this is made much worse by the fact that `Display` impls are inherently lossy, but the name `to_string` does not imply this.

**Comment by tmandry** [comment]
From https://github.com/rust-lang/libs-team/issues/550#issuecomment-2897075246:

> [@tmandry](https://github.com/tmandry) The entire point of ByteStr is to be "conventionally UTF-8". The type exists to serve that role. It is already widely used in the ecosystem as "bstr::BStr", including the behavior of Display and Debug. (Also, the right place for ByteStr discussion is on its own tracking issue.)

I get that the point of ByteStr is to be conventionally UTF-8; my point is I don't think the name does a good job of conveying this role.

> I agree that `to_string` is not a great name. That should never have been based on Display. But that's a universal problem, and would also apply if someone called `to_string` on some other type that didn't convert losslessly to a string.

Yes, and its existence has caused other string types not to implement Display. Given that we have `to_string` when implementing Display, I think we should strive to make it as clear as possible when a user is going down a potentially-lossy path.

**Comment by jhpratt** [comment]
One nit: if the `ByteStr` naming is going to stick around instead of `BStr`, should the module still be named `bstr`? That seems unusual.

**Comment by clarfonthey** [comment]
That's very fair; I'm mostly commenting from the perspective of both having used `BStr` in an API before and then later abandoning it for the reasons I mentioned, and the fact that `[ascii::Char]` might be better suited in places where `BStr` would otherwise be used.

You do bring up a good point about non-std implementations, and the `Debug` issues.

**Comment by kupiakos** [comment]
> we have `to_string` when implementing Display, I think we should strive to make it as clear as possible when a user is going down a potentially-lossy path

I agree - the `to_string` lossiness issue is too ambiguous for the standard library. A `display` method and _no `impl Display`_ is consistent with `Path`/`OsStr` (though it's too late to add `lossy` there). If we add `fn to_string_lossy(&self) -> Cow<'_, str>` to `ByteStr` as the inverse of `String::from_utf8_lossy`, then the most well-lit path to convert `ByteStr` to `String` is both likely-zero-copy and mentions the data loss.

**Comment by clarfonthey** [comment]
What is the current policy for recommending extensions to these types in their current form? A new ACP? Or just pull requests to improve things?

One massive issue with the current behaviour is the lack of any form of `DerefMove`, which means that any methods on `Vec` which consume the vector entirely do not work via `ByteString`. For example, calling `into_boxed_slice` on a `ByteString` to convert it into a `Box<[u8]>` will just throw an error. rustdoc is smart enough to not show these methods under the `Deref` section on `ByteString`, but users may still want to call them.

Another issue is that several methods that you might call via `ByteStr` subtly return bytes again when you probably won't want them to. For example, `to_ascii_uppercase` returns `Vec<u8>` even though it feels most apt to return `ByteString`.

Also, there are a bunch of methods that should *probably* be added for parity with regular `String`, like `into_boxed_byte_string` on `Box<ByteStr>`.

**Comment by AhoyISki** [comment]
Are there plans to add a sort of `TryChars` iterator over a `ByteStr`? Something that iterates through the `&[u8]` and returns values like:

```rust
enum BikeshedableTryChar {
    Valid(char)
    Invalid(u8)
}
```

Because `Utf8Chunks` doesn't cut it for really large pieces of text that are almost entirely utf8, since you'd have to reparse the whole thing, even if you wanted to get just a few `char`s in a specific range.

**Comment by AhoyISki** [comment]
I meant more that you'd have to iterate over massive parts of the `ByteStr` in order to figure out where the `Utf8Chunk` lies, which you may not want to do.
Also, it doesn't necessarily have to return the bytes, it could return replacement characters, or the length of the invalid region, I'm just giving a suggestion for a feature that would make certain operation more efficient.

**Comment by clarfonthey** [comment]
To be clear, I am mostly trying to hone in on what operations you're trying to make more efficient, and to clarify that pulling chars out of UTF-8 chunks is not necessarily less efficient than pulling out chars-or-bytes.

Like, would you like a `utf8_chunk_indices` iterator, similar to `char_indices`? Would you like `utf8_chunks` to work differently somehow? That's what I'm mostly wondering about.

Regardless, I'm not sure that the `ByteStr` issue is the right place to discuss this. Probably, it might be worth filing an ACP with the types of problems you're trying to solve and a method could be added orthogonal to it, similarly to how `utf8_chunks` is stable now separate from that API.

**Comment by ChrisDenton** [comment]
> The longstanding policy for std was that types which can't round trip ToString -> FromStr without changing their contents do not implement `Display`

To be clear that policy was already (accidentally?) broken for `SocketAddrV6`, hence the added equivocation. No doubt other types don't exactly round trip (I think floats have had issues with this in the past, though arguably that's a bug).

**Comment by DanielEScherzer** [comment]
Is there something I can do to help move this towards stabilization? I'm working on FFI interaction with PHP strings and those are basically just `[u8]` slices, so I've started using the bstr crate, but having this built in would be really helpful

**Comment by tgross35** [comment]
> Is there something I can do to help move this towards stabilization? I'm working on FFI interaction with PHP strings and those are basically just `[u8]` slices, so I've started using the bstr crate, but having this built in would be really helpful

Perhaps you could open some discussion on Zulip? At a minimum, it seems like the unresolved things are:

1. Do we want https://github.com/rust-lang/rust/pull/146219 or the current version
2. Naming, both of the types and of the module
3. Should `Display` use the replacement character or escape? Seems like consensus may be heading toward the former.
4. We need an `Index` impl for the new `std::range` types. You could probably open a PR if interested.
5. In general, somebody probably needs to look over every trait and methods on `&[u8]` and `Vec<u8>` to see if there are any that should be implemented on `ByteStr`/`ByteString` with a different signature. (To avoid https://github.com/BurntSushi/bstr/issues/190).

I think it would be valuable to hear @BurntSushi's thoughts as well, as the author of the `bstr` crate.

**Comment by BurntSushi** [comment]
Thank you for the ping. I'm on vacation now, but once I'm back, I'll share some thoughts and opinions.

**Comment by clarfonthey** [comment]
Also to reshare the justification I wrote up for making the type opaque: https://github.com/rust-lang/rust/pull/146219

> The main issue with `ByteStr` is that it's not a "real" string type: they're just wrappers over `[u8]` and `Vec<u8>` with barely any methods of their own, relying instead on `Deref` impls. The inability to move out of deref here prevents methods like `Vec::leak` from being used without much effort, and the fact that the two types aren't linked via `Deref` means that any extra methods added to `ByteStr` would have to be added to `ByteString` too.
> 
> The new wrapper types act more in line with a "real" string type while still allowing easy access to the underlying `[u8]` and `Vec<u8>` without any complaining. In particular, this means that `ByteString` has an `as_mut_vec` method that can be executed safely (unlike `String::as_mut_vec`) and `ByteStr` has an `as_bytes_mut` method that can be executed safely. Any methods added to `ByteStr` can be accessed directly by `ByteString` due to the deref implementation, avoiding the need to add new methods to both types.
> 
> Note that these changes parallel `OsStr` and `OsString` for a reason: I was hoping to update `OsStr` and `OsString` to internally use these types on non-Windows targets. This feels like an obvious case where these types would be useful, and considering how I totally failed due to the issues I mentioned, I decided to at least start with what would be necessary to achieve that and we can build from there.

**Comment by tgross35** [comment]
Couldn't we add a `leak` method and similar to `ByteString` if that is valuable? `ByteStr::new(some_bytestring.0.leak())` also doesn't seem all that bad, so perhaps I'm missing something.

It also seems like a `ByteString::as_byte_str` or `<[u8]>::as_byte_str` (like `bstr::ByteSlice`) would smooth the lack of `Deref`. Or at least, that's about an equal speedbump to needing `as_bytes`/`as_mut_vec` to get slice/vec methods with the proposal.

That said, having `ByteString` deref to `ByteStr` does seem nicely consistent with other string types. It also avoids the problem of needing to be extra careful about current and future `[u8]` methods and traits that we won't be able to override for `ByteStr`.

**Comment by BurntSushi** [comment]
> Should Display use the replacement character or escape? Seems like
> consensus may be heading toward the former.

Has there been discussion about this? If so, I've missed it. I think the
replacement codepoint is the right thing to do here. It's not just about
`Display` output, but
[what we do for lossy decoding](https://doc.rust-lang.org/nightly/std/str/struct.Utf8Chunk.html#method.invalid)
in general. Moreover, these are meant to be types that represent _text_, not
binary data. The `Display` impl should reflect that textual interpretation.

> 1. Do we want [Make ByteStr/ByteString a more-opaque wrapper like
> OsStr/OsString #146219](https://github.com/rust-lang/rust/pull/146219) or
> the current version

I'm strongly opposed to this. As far as I can tell, this is reverting to the
[`bstr 0.1` design](https://docs.rs/bstr/0.1.4/bstr/), which I [tossed in
the trashbin](https://github.com/BurntSushi/bstr/issues/5) after I tried to
actually use it. The fundamental problem is exactly the issue you're saying is
problemtic with the current design: you end up in conversion hell. That is, to
get access to "string" methods, you need to get a `ByteStr`. But then to pass
it to anything that wants a `&[u8]`, you need to convert it back. (And you
can't idiomatically use `Deref` because your encapsulated `ByteStr` type has a
whole bunch of methods on it.)

> Note that these changes parallel OsStr and OsString for a reason: I was
> hoping to update OsStr and OsString to internally use these types on
> non-Windows targets. This feels like an obvious case where these types
> would be useful, and considering how I totally failed due to the issues
> I mentioned, I decided to at least start with what would be necessary to
> achieve that and we can build from there.

@clarfonthey I want to strongly emphasize something that I learned in the course of
designing and building `bstr`: **std already has a byte string type and
its name is `&[u8]`.** We are stuck with that, for better or worse. `&[u8]` is
already used in way too many places. So IMO, any design (like an encapsulated
`ByteStr`) that tries to push it aside to become *the* byte string type ends
up being too annoying to use because everything already accepts or provides
`&[u8]`.

In other words, `bstr` is two things.

Firstly, `bstr` beefs up the existing byte string type (named `&[u8]`)
with string-like methods for use cases where a `&[u8]` is **intended to be
conventionally UTF-8, but not required to be so.** A perfect example is the
`&[u8]` used in `std::io::Read`. This is ripgrep's model of the world, for
example. (And it sometimes goes to great effort to uphold this model, for
example, by automatically transcoding UTF-16 to UTF-8 before ripgrep's core
ever sees it.) ripgrep rarely ever works with `&str`. It's `&[u8]` all the way
down.

Secondly, `bstr` provides separate types for byte strings. But why do that if
I've claimed that `&[u8]` is already the standard byte string type? One (but
not the only) especially compelling reason is because `&[u8]` cannot be used
as a target of trait implementations where `&[T]` is also a target because
of how Rust's trait system works. (I don't even know if specialization would
fix this use case, but at least conceptually in the limit it would I think.)
And analogously for `Vec<u8>` and `bstr::BString`. The most *critical* trait
implementation for `bstr::BStr` is, IMO, the `core::fmt::Debug` implementation.
It is a *significant* quality of life improvement to be able to treat `&[u8]`
as a conventionally UTF-8 string for debugging code.

I write `Debug` and `Display` adapter trait implementations for `&[u8]` in
several of my crates (this can be made simpler with `[u8]::utf8_chunks` once
MSRVs are sufficiently bumped):

* https://github.com/BurntSushi/jiff/blob/66f6ec2d0b75112deadf7e067dbf2ef114d41c50/src/util/escape.rs
* https://github.com/rust-lang/regex/blob/7b96fdc9d5fe6a0cb4efe30e6689b050493fc1e1/regex-automata/src/util/escape.rs
* https://github.com/rust-lang/regex/blob/7b96fdc9d5fe6a0cb4efe30e6689b050493fc1e1/regex-syntax/src/debug.rs

It's exactly the kind of thing that ought to live in `core` IMO because it's
too small to warrant bringing in a crate dedicated to it. But it's also
annoying enough to get right that it really should be done somewhere once.

Something like `serde` I imagine would provide `Serialize` and `Deserialize`
trait implementations for this type as well.

The other reason why `bstr::BStr` and `bstr::BString` exist *aside* from being
targets of trait implementations is because they interact really well with
existing `derive(Debug)` support. You put a `BStr` in your type, `derive(Debug)`
on it, and all of a sudden you get effectively exactly what you want in the
output instead of a vector of integers. Not only that, but it *describes the
intent*. It's an elegant solution to this problem IMO given the status quo.

Now, what about your difficultly in using `std::bstr::ByteStr` in `OsStr`?
I'm not sure I quite follow it there, but IMO, it would be perfectly reasonable
to just use `[u8]` there internally. And any string stuff you want to do should
be methods on `[u8]`. I think @joshtriplett mentioned somewhere that this was
just the beginning. I'd expect more methods on `[u8]` (like `bstr::ByteSlice`
provides) to significantly ease the burden here.

To get a good idea of how `bstr::BStr` is used in a real project, I'd suggest
looking at [`bttf`](https://github.com/BurntSushi/bttf). It was honestly quite
a pleasure and it made debugging lovely.

OK, to quote something I said above, because it's worth highlighting
and elaborating on:

> The most *critical* trait implementation for `bstr::BStr` is, IMO, the
> `core::fmt::Debug` implementation. It is a *significant* quality of life
> improvement to be able to treat `&[u8]` as a conventionally UTF-8 string.

I was quite careful to omit `Display` here. `Debug` is useful because it is
derivable. I think the `Display` impl is less useful, and I think that the
`to_string()` problem is an unfortunate footgun that won't be able to be
realistically stabilized (for good reason). To be more explicit about this,
I'll respond to @withoutboats commentary on this:

> My own personal opinion is that this policy is not good, that this
> should not be a contract on Display, and that Path and PathBuf
> should just implement Display. But I was overruled when I was on the
> libs team and it was held to be a backward compatibility commitment
> that Rust std had made at 1.0. Before stabilizing this, I think
> the libs team should either commit to changing this policy (and by
> implication, add Display to the path types) or remove the Display
> impl from these types and add a display helper like the path types
> have.

IMO, this is an incomplete characterization of the referenced #49868
issue. It was not just backcompat legalese.
[I myself was strongly in favor of implementing `Display` on `Path`](https://github.com/rust-lang/rust/pull/49868#issuecomment-380409024),
but that opinion has [weakened over time](https://github.com/rust-lang/rust/pull/49868#issuecomment-380896752)
once [the problem with `to_string()` was raised](https://github.com/rust-lang/rust/pull/49868#issuecomment-380499661).
IMO, this is a far stronger argument against implementing `Display` in a way that is lossy because
a `to_string()` method call strongly indicates a non-lossy transformation.
I realize `bstr` does this, but I'm not sure it's really pulling its weight
because **std doesn't provide facilities for deriving `Display`**. And even if
and when it does (I hope it does), it will need to accommodate `Path` which
doesn't implement `Display`. I think `anyhow` and other crates that provide a
way to `derive(Display)` already solve this problem in a reasonably satifactory
way I think. The main bummer in my experience is writing ad hoc error messages
using a `Path`. It's annoying to put a `path.display()` everywhere. And
probably would be as well for `ByteStr`.


**My point here is that we should probably just drop the `Display`
implementation on `ByteStr` and `ByteString`, but keep the `Debug`
implementation. The latter is where the real goods are I think. Then we can add
a `[u8]::display()` routine analogus to `Path::display`.**

The alternative, [as suggested by @tmandry](https://github.com/rust-lang/rust/issues/134915#issuecomment-2899431966),
is to use names other than `ByteStr`/`ByteString` that more accurately describes the
assertion it is making ("conventional UTF-8"). I think this is probably not a
resolvable alternative in a way that makes everyone happy enough to allow these
types to be stabilized. So we are really left with two choices I think:

1. Accept the `to_string()` lossy footgun. (Also probably not able to be stabilized.)
2. Drop the `Display` impls on `ByteStr` and `ByteString`. (And probably also
   `FromStr` on `ByteString`. It was probably a gaffe of mine to add that to
   `bstr` in the first place.)

I think we should do (2).

**Comment by DanielEScherzer** [comment]
> Drop the `Display` impls on `ByteStr` and `ByteString`. (And probably also `FromStr` on `ByteString`. It was probably a gaffe of mine to add that to `bstr` in the first place.)

I sent #158215 for the `Display` removals, will follow it up with a `FromStr` removal afterwards

**Comment by DanielEScherzer** [comment]
Hmm, the display is used for #139984

```
impl CStr {
    pub fn display(&self) -> impl fmt::Display {
        crate::bstr::ByteStr::from_bytes(self.to_bytes())
    }
}
```

We could
* inline/move the ByteStr display logic so that it can be used for CStr but not ByteStr
* remove the CStr::display() implementation
* make the CStr::display() match the debug() version, where invalid UTF-8 is shown as hex, and then it should be as simple as `format!("{:?}", crate::bstr::ByteStr::from_bytes(self.to_bytes()))` to return a string with the display contents

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/libs-api meeting. In particular, we talked at length about `Display`.

We felt that because `ByteStr` is *conventionally UTF-8*, it makes sense to have a `Display` impl. We don't want exclusively `Debug`; we want people to be able to display a human-readable representation.

Our main concern was not actually with `Display` itself; it was with `to_string()`, which is automatically provided for anything implementing `Display`. And we don't want people to be able to easily call `.to_string()` and have that be a lossy method, when they should call `.to_string_lossy()` or call something that returns a `Result`.

Some folks also had concerns about `FromStr`.

But we don't want to let those stop us from shipping a useful `Display`, because people want to use `ByteStr` as a type for data that comes from the user and goes to the user; that's one of its common use cases.

So, our conclusion was that we should provide `Display`, defer `FromStr` for now, and provide an inherent `to_string()` method that is marked as `#[deprecated]` with a deprecation method steering people towards either `to_string_lossy()` or to a method like `ByteStr::into_string` (which we haven't added yet) that checks for non-UTF-8 and returns `Result`. That inherent method will then shadow the one from the `ToString` trait.

**Comment by DanielEScherzer** [comment]
I'll send a patch to point to `to_string_lossy()` for now

Edit: there isn't a `to_string_lossy()` that appears available, unless I'm missing something I'll point to `str::from_utf8()` / `String::from_utf8_lossy()`

**Comment by joshtriplett** [comment]
Shall we stabilize `ByteStr` and `ByteString`, as described?

@rfcbot merge libs-api

One concern we have to settle before doing so: should we keep the `Deref`s to `[u8]`/`Vec<u8>`?

If we keep them, then adding new methods directly to `ByteSr`/`ByteString` becomes a breaking change, and we'd be committed to doing that via trait (the way the `bstr` crate does).

If we drop the `Deref` impls, then we lose some convenience of usage, but we gain the ability to add methods to `ByteStr`/`ByteString` directly.

@rfcbot concern deref-vs-methods

**Comment by rust-rfcbot** [comment]
@joshtriplett has proposed to merge this. The next step is review by the rest of the tagged team members:

* [ ] @Amanieu
* [x] @BurntSushi
* [x] @dtolnay
* [x] @joshtriplett
* [ ] @the8472

Concerns:

* deref-vs-methods (https://github.com/rust-lang/rust/issues/134915#issuecomment-4971877039)

Once a majority of reviewers approve (and at most 2 approvals are outstanding), this will enter its final comment period. If you spot a major issue that hasn't been raised at any point in this process, please speak up!

cc @rust-lang/libs-ping: FCP proposed for libs, please feel free to register concerns.
See [this document](https://github.com/rust-lang/rfcbot-rs/blob/master/README.md) for info about what commands tagged team members can give me.

**Comment by clarfonthey** [comment]
To clarify on the `Deref` issue: I think that the choices made in `bstr`'s early development don't necessarily reflect the current state of Rust. Looking at the version history from what I can see, these were made in 2019 (Rust version at release of bstr=0.2.0 was 1.39.0), and since then we've added more conversions for working with the various stdlib string types.

* `PathBuf::as_mut_os_string` (1.70.0 / 2023)
* `OsStr::as_encoded_bytes` (1.74.0 / 2023)
* `OsStr::display` (1.83.0 / 2024)

And actually, converting between `Path`, and `OsStr`, and `[u8]` is actually a bit of a mess at the moment, but not *really* that big of a deal, since you can just do:

```rust
fn uses_os_str(s: &OsStr) {
    let s = s.as_encoded_bytes();
    // ...
}
```

And all of a sudden, everything is fine. For `ByteStr`, similarly, I don't really think it's that bad to have to call `as_bytes` or `ByteStr::new` frequently to do stuff. We already do a lot of this with `OsStr` and `Path` and we've already added a lot of methods to avoid having to do conversions entirely.

Like, I don't really see much of an issue with most of the use cases for `ByteString` and `ByteStr` being covered by the same handful of methods we have on `OsString`, plus an `as_mut_bytes` etc. method which doesn't even require an unsafe block. (Unlike `OsStr::from_encoded_bytes`, which does.)

Also, the example I included in the PR offering an opaque type was particularly compelling: `Vec::leak` will not work directly on `ByteString`, and you would have to do `ByteString::leak` instead, similar to anything that might take the vector by-value. So, you would need an `into_vec` method anyway, or to just do `.0` if we make the fields public.

The lack of deref makes a lot of APIs substantially worse, e.g. not being able to call `get(bstr)` on a `HashMap` with `ByteString` keys. I would encourage looking at the specific API offered by #146219 since it feels like a majority of cases would be covered by it and conversions wouldn't really be necessary ever.

**Comment by BurntSushi** [comment]
@clarfonthey Thanks for the pushback! I talked with @joshtriplett a bit and I think I'm becoming a bit more sympathetic to your view here. In particular, one thing that I think would be exceptionally difficult in a world where `ByteStr` has a deref impl is the naming of methods on `[u8]` for stringly operations. I think it would be _exceptionally_ difficult, actually, to make room for such methods on `[u8]` where it also has to contend with methods on `[T]`. A proper newtype gives that more space to breathe.

I do think that one thing std has going for it versus [`bstr 0.1`](https://docs.rs/bstr/0.1) is that we can add a `[u8]::as_bstr()` method. That was harder in `bstr` because it required an extension trait _plus_ the newtypes. But maybe that makes the difference.

I looked at `bttf`, and it currently has 41 instances of `.as_bstr()`, which is pretty hefty already. It's not like `bstr 1.0`'s design is doing _that well_ on that dimension. Now, that doesn't tell me how many of those routines I'd need in a world where `ByteStr` doesn't have deref, though. Because `bstr` provides methods directly on `[u8]`. But I don't have an easy way of checking that.

> The lack of deref makes a lot of APIs substantially worse, e.g. not being able to call `get(bstr)` on a `HashMap` with `ByteString` keys.

I am a little confused here. I think the deref impl in question is `ByteStr -> [u8]`. But we probably always want a deref impl on `ByteString`. If we have a deref impl on `ByteStr`, then that would imply a deref impl of `ByteString -> Vec<u8>` (which is the status quo). If we dropped the deref impl on `ByteStr` and committed to putting stringly methods on `ByteStr`, then we'd have a deref impl of `ByteString -> ByteStr`, right? So in either world, `get(ByteStr)` on `HashMap<ByteString, ...>` would work?

(The different deref impls on `ByteString` is also a good reason why this is indeed something we have to decide now before stabilization. Unless we leave off all deref impls and all methods, I think.)

Something I noticed in #146219:

> You may have also noticed these omitted, which I'm pointing out separately due to a bit of an API complication. CStr and CString decide to provide full access to the UTF-8 error here, but OsStr and OsString do not. I decided to go with OsStr and OsString's API because they were simpler, but I personally think that the decision on what to do here should also be paired with one on what to do with the existing API discrepancy as well:

IMO, I feel strongly that we shouldn't repeat the mistake of `OsStr`/`OsString` here. We should optimize for `?` like `CStr`/`CString` do and provide a real error type with a nice error message. The error type can expose the allocation in the case where that is useful.

**Comment by joshtriplett** [comment]
@BurntSushi wrote:
> If we dropped the deref impl on `ByteStr` and committed to putting stringly methods on `ByteStr`, then we'd have a deref impl of `ByteString -> ByteStr`, right?

Yes, when we talked about dropping `Deref` it was about dropping the `Deref` to `Vec<u8>`/`[u8]`; we'd still want one from `ByteString` to `ByteStr`.

I wrote:
> So, our conclusion was that we should provide `Display`, defer `FromStr` for now, and provide an inherent `to_string()` method that is marked as `#[deprecated]` with a deprecation method steering people towards either `to_string_lossy()` or to a method like `ByteStr::into_string` (which we haven't added yet) that checks for non-UTF-8 and returns `Result`. That inherent method will then shadow the one from the `ToString` trait.

One outcome of the conversation with @BurntSushi: I realized that there's a better approach here. Rather than adding a `to_string` method that shadows `ToString::to_string` and is immediately deprecated, we can just directly provide a `to_string` method that *is* the conversion returning `Result<String, AppropriateErrorType>`, no deprecation needed.

cc @DanielEScherzer

**Comment by camsteffen** [comment]
Maybe we should ship a `lossy_to_string` lint or `lossy_display_usage` lint? I do sympathize with the NOT A CONTRIBUTION that we should try to land on a consistent policy for Display, and a lint could help to paint that picture? I guess the inherent `to_string` resolves this well enough, but I don't really understand why `to_string()` is more concerning than other usage of Display.

**Comment by BurntSushi** [comment]
> but I don't really understand why `to_string()` is more concerning than other usage of Display.

I think [@tmandry's comment above](https://github.com/rust-lang/rust/issues/134915#issuecomment-2899431966) captures my own sentiments pretty well.

**Comment by clarfonthey** [comment]
FWIW I realised I was just totally wrong earlier and that `HashMap`'s use of `Borrow` makes `Deref` less of an issue in most cases, but having the `ToOwned` and `Deref` impls not agree feels like a pretty annoying issue. It also means that `Cow` will not do what people expect, for example.

To be clear, I am 100% flexible with what API we use for the opaque wrapper and agree that `OsStr` is probably not the model citizen in terms of what the best decision is, and my main issue is whether `Deref` binds `ByteStr` <-> `ByteString` instead of `ByteStr` <-> `[u8]` and `ByteString` <-> `Vec<u8>`.

I also would strongly like to see `ByteStr` effectively become the wrapper underlying `OsStr` for Unix barring some weird technicalities that we haven't thought of.

**Comment by BurntSushi** [comment]
> and my main issue is whether `Deref` binds `ByteStr` <-> `ByteString` instead of `ByteStr` <-> `[u8]` and `ByteString` <-> `Vec<u8>`.

Yes, I think I'm leaning toward the former. If nothing else because of the naming hell that would be required if we go for the latter. Fitting a bunch of stringly methods on `&[u8]` seems very very difficult. But we can side-step all of that with `[u8]::as_byte_str()`.

I think having this in std will give us a better story around conversions. `bstr 0.1` didn't have a `[u8]::as_bstr()` method. I think that will help, although I do think that something like `slice.as_byte_str().split("needle").map(|b| b.as_bytes())` will not be terribly uncommon. i.e., "switch to byte str to do string things and then switch back to `&[u8]` because that's what something requires downstream."

In the case where we have `Deref` that binds `ByteStr` to `[u8]`, then you can just do stringly things on `[u8]` without caring about using a separate type. _But_, when you do have a `ByteStr` (e.g., declared in your types because it's convenient), then you're still needing to convert there too. So there really isn't a way out of conversion hell I think. We have to pick something and I think having `ByteString` deref to `ByteStr` is actually probably the least worst option. And yes, that means our stringly routines go on `ByteStr`.

**Comment by clarfonthey** [comment]
Yeah, it is definitely worth calling out that the ability to just add methods to all the primitive types for conversion makes this substantially more appealing for libstd.

I also don't think the API surface has to be particularly large on the `ByteStr(ing)` types to cover most of the desired use cases. In general, if you're going to be doing any super fancy conversions you probably just want a string, since `ByteStr`'s model of "want a string but degrade to verbatim bytes if we can't" doesn't really leave a lot of good meanings for those conversions on non-Unicode data. There will be some conversions that make sense, like case folding, but then for stuff like normalisation, only NFD makes sense, etc.

One other compelling bit is that `OsString`'s opacity for WTF-8 allows it to have a flag that marks UTF-8 validity ahead of time make operations easier to perform, and something like that for `ByteString` could be particularly useful to have. We could even potentially include that flag on `ByteStr` with some custom DST metadata shenanigans in the future.

**Comment by BurntSushi** [comment]
> I also don't think the API surface has to be particularly large on the `ByteStr(ing)` types to cover most of the desired use cases. In general, if you're going to be doing any super fancy conversions you probably just want a string, since `ByteStr`'s model of "want a string but degrade to verbatim bytes if we can't" doesn't really leave a lot of good meanings for those conversions on non-Unicode data.

I would generally expect it to grow to the size of `str` personally. For non-Unicode data on Unicode operations, you can use the replacement codepoint. That's how Go's standard library works and also how `bstr` works. I had always wanted to add normalization to `bstr`, and I think that works by just substituting in the replacement codepoint for invalid UTF-8. But `str` doesn't even have normalization in std. Anyway, using `U+FFFD` is how `bstr`'s `chars()`, `graphemes()` and `sentences()` iterators work. _And_ because of that, you actually get iterators of `&str` and not byte strings. Which is kinda nice.

**Comment by thaliaarchi** [comment]
> One other compelling bit is that `OsString`'s opacity for WTF-8 allows it to have a flag that marks UTF-8 validity ahead of time make operations easier to perform, and something like that for `ByteString` could be particularly useful to have. We could even potentially include that flag on `ByteStr` with some custom DST metadata shenanigans in the future.

For what it's worth, having worked on `OsString`, I'm not particularly fond of that flag. I'd be happy with it, if it took no extra space, but it brings the size of `OsString` on Windows from 24 to 32 bytes. I've idly considered using the MSB of len or capacity for `is_known_utf8: bool`, but then every operation would need to reconstruct the Vec from its parts.

**Comment by BurntSushi** [comment]
I'd be pretty skeptical of it too, but I'm trying to refrain from getting into the weeds of implementation details. :-) I think the important bit here is that we could at least _consider_ it. Which is fair, but I'm not sure it's a good idea.

**Comment by clarfonthey** [comment]
To unblock FCP, I think it would be fair to accept the following API as uncontested (and also this kind of doubles as a stabilisation report):

```rust
// core/alloc respectively; note lack of repr(transparent) or inner field being public API to avoid temptation to use `.0`
pub struct ByteStr([u8]);
pub struct ByteString(Vec<u8>);

// note: for now, these would be canonical conversions, due to inference failures on various traits
impl [u8] {
    pub const fn as_byte_str(&self) -> &ByteStr;
    pub const fn as_mut_byte_str(&mut self) -> &mut ByteStr;
    pub fn into_boxed_byte_str(self: Box<[u8]>) -> Box<ByteStr>;
}
impl Vec<u8> {
    pub fn into_byte_string(self) -> ByteString;
}
impl ByteStr {
    // mirrors `str` methods, but more safe due to invariant differences
    pub const fn as_bytes(&self) -> &[u8];
    pub const fn as_bytes_mut(&mut self) -> &mut [u8];
    pub fn to_byte_string(&self) -> ByteString;
    pub fn into_byte_string(self: Box<ByteStr>) -> ByteString;
    pub fn into_boxed_bytes(self: Box<ByteStr>) -> Box<[u8]>;
}
impl ByteString {
    // mirrors `OsString` methods
    pub fn as_byte_str(&self) -> &ByteStr; 
    pub fn as_mut_byte_str(&mut self) -> &mut ByteStr;
    pub fn leak<'a>(self) -> &'a mut ByteStr;

    // mirrors `String` methods
    pub fn as_mut_vec(&mut self) -> &mut Vec<u8>;
    pub fn into_bytes(self) -> Vec<u8>;
    pub fn into_boxed_byte_str(self) -> Box<ByteStr>;
}
impl Box<ByteStr> {
    // mirrors `Box<str>` method
    pub fn into_boxed_bytes(self) -> Box<[u8]>;
}

// possible extensions include direct methods from `String` and `str`, but for now you can just deal with intermediate method calls

// include all the same indexing that `[u8]` and `Vec<u8>` can do

// note the ones omitted due to inference failures; assume `Mut` implies `Ref`
impl AsMut<ByteStr> for ByteStr;
impl AsMut<ByteStr> for ByteString;
impl AsRef<ByteStr> for str;
// impl AsMut<ByteStr> for [u8];
impl AsMut<[u8]> for ByteStr;
impl AsMut<[u8]> for ByteString;
impl BorrowMut<[u8]> for ByteStr;
impl BorrowMut<[u8]> for ByteString;
// impl BorrowMut<ByteStr> for [u8];
// impl BorrowMut<ByteStr> for Vec<u8>;
// impl Borrow<ByteStr> for str;
// impl Borrow<ByteStr> for String;

// note the ones omitted due to inference failures
// impl From<&[u8; N]> for &ByteStr;
// impl From<&[u8]> for &ByteStr;
// impl From<&str> for &ByteStr;
// impl From<&[u8; N]> for ByteString;
// impl From<[u8; N]> for ByteString;
// impl From<&[u8]> for ByteString;
// impl From<Vec<u8>> for ByteString;
// impl From<&str> for &ByteStr;
// impl From<&str> for ByteString;
// impl From<String> for ByteString;
// impl From<&ByteStr> for &[u8];
// impl From<&mut ByteStr> for &mut [u8];
impl From<ByteString> for Vec<u8>;
impl From<&ByteStr> for ByteString;
impl From<&ByteStr> for Cow<'_, ByteStr>;
impl From<ByteString> for Cow<'_, ByteStr>;
impl From<Box<[u8]>> for Box<ByteStr>; // also Rc, Arc
impl From<Box<ByteStr>> for Box<[u8]>; // also Rc, Arc
impl TryFrom<&ByteStr> for &str;
impl TryFrom<&ByteString> for &str;
impl TryFrom<&mut ByteStr> for &mut str;
impl TryFrom<&ByteStr> for String;
impl TryFrom<ByteString> for String;

// comparisons; note that PartialOrd implies PartialEq but only PartialEq indicates inference failure
// reverse comparisons implemented too, and ByteStr on RHS implies also ByteString
impl PartialEq<[u8]> for ByteStr;
impl PartialEq<str> for ByteStr;
impl PartialEq<&[u8]> for ByteStr;
impl PartialEq<&str> for ByteStr;
impl PartialEq<[u8; N]> for ByteStr;
impl PartialEq<&[u8; N]> for ByteStr;
impl PartialEq<Vec<u8>> for ByteStr;
impl PartialEq<String> for ByteStr;
impl PartialOrd<Cow<'_, [u8]>> for ByteStr;
impl PartialOrd<Cow<'_, str>> for ByteStr;
impl PartialOrd<Cow<'_, ByteStr>> for ByteStr;
impl PartialOrd<ByteStr> for ByteString;
impl PartialOrd<&ByteStr> for ByteString;

// all the `FromIterator` impls for bytes, slices, strings, byte strings work as expected

impl {Debug, Ord, Hash, Default} for {ByteStr, ByteString};
impl ToOwned<Owned = ByteString> for ByteStr;
impl DerefMut<Target = ByteStr> for ByteString;
impl Clone for Box<ByteStr>;
```

There is some point of contention over `Display` that I think we're fine getting over:

```rust
impl Display for {ByteStr, ByteString};
```

The below methods were proposed for the opaque wrapper which should be okay in an initial API:

```rust
impl ByteStr {
    // mirrors `OsStr` methods
    pub const fn is_empty(&self) -> bool;
    pub const fn len(&self) -> usize;
    pub const fn is_ascii(&self) -> bool;
    pub const fn make_ascii_lowercase(&mut self);
    pub const fn make_ascii_uppercase(&mut self);
    pub fn to_string_lossy(&self) -> Cow<'_, str>;
    pub fn to_ascii_lowercase(&self) -> ByteString;
    pub fn to_ascii_uppercase(&self) -> ByteString;
}

impl ByteString {
    // mirrors `Vec` methods on `OsString`
    pub const fn new() -> ByteString;
    pub fn with_capacity(capacity: usize) -> ByteString;
    pub fn clear(&mut self);
    pub fn capacity(&self) -> usize;
    pub fn reserve(&mut self, additional: usize);
    pub fn try_reserve(&mut self, additional: usize) -> Result<(), TryReserveError>;
    pub fn reserve_exact(&mut self, additional: usize);
    pub fn try_reserve_exact(&mut self, additional: usize) -> Result<(), TryReserveError>;
    pub fn shrink_to_fit(&mut self);
    pub fn shrink_to(&mut self, min_capacity: usize);

    // mirrors `String` methods
    pub fn truncate(&mut self, len: usize);
}
```

Note that this excludes a few key methods which I think should be revisited due to the various inference failures on `AsRef`, in favour of other methods in the meantime (intermediate conversions with bytes):

```rust
impl ByteStr {
    // constructors based on `OsStr`, but with new mutable version
    pub const fn new<S: ?Sized + [const] AsRef<ByteStr>>(bytes: &S) -> &ByteStr;
    pub const fn new_mut<S: ?Sized + [const] AsMut<ByteStr>>(bytes: &mut S) -> &mut ByteStr;

    // mirrors `OsStr` methods
    pub const fn eq_ignore_ascii_case<S: [const] AsRef<ByteStr> + [const] Destruct>(&self, other: S) -> bool;
}

impl ByteString {
    // this parallels `OsString`, but is opposite to `String`, which has `push` and `push_str`
    pub fn push<S: AsRef<ByteStr>>(&mut self, s: S);
    pub fn push_byte(&mut self, b: u8);
}
```

The main issue is that to make these match `OsStr`, we would want these to be `AsRef<ByteStr>`, but since that doesn't work properly, we would need `AsRef<[u8]>` instead. (status quo)

My proposal is to hold off on these for now because the inference failures substantially reduce the ergonomics of a lot of the conversions and these are not so vitally important that we need to resolve this right now. I would rather stabilise most of the proposed API rather than deal with this right now, and hey, maybe by some miracle the new trait solver fixes everything. (lmao)

-----

Also for an explicit clarification: if we're fine with this, I'd basically just rebase my opaque wrapper PR and the stabilisation PR would stabilise the proposed subset of that.

**Comment by BurntSushi** [comment]
@clarfonthey Thank you so much for putting this together! This is looking great to me.

I agree with holding off on things that want `AsRef<ByteStr>`. That seems okay. We'll want to figure that out sooner rather than later though, since I think we'll want that for a ton of stringly APIs on `ByteStr`. (e.g., Substring search.) `AsRef<[u8]>` is not a terrible compromise though if that's our only option.

One thing I don't think I see here is an inherent `to_string()` method on `ByteStr` that returns `Result<&str, AppropriateErrorType>` and `ByteString` that returns `Result<String, AppropriateErrorTypeThatIsNotJustByteString>`. Those are meant to shadow the `ToString` methods. I think we need this if we want to `impl Display for {ByteStr, ByteString}`.

We will also want documentation on `ByteStr` and `ByteString` warning folks about the `to_string()` shadowing _and_ how that might not help in all cases, e.g., passing `ByteStr`/`ByteString` to something that wants an `impl Display`.

**Comment by clarfonthey** [comment]
Yeah, the string conversion methods were excluded because I forgot to copy-paste those since they were "disputed," but I agree we should just copy the `CStr` API there.

**Comment by camsteffen** [comment]
(digressing about Display)

> We felt that because `ByteStr` is *conventionally UTF-8*, it makes sense to have a `Display` impl.

I don't disagree with this direction but also I don't think that this differentiates `ByteStr` from `Path`. The reason that `Path` doesn't implement `Display` isn't because it's not (conventionally) UTF-8. It's because the *conversion* to UTF-8 is potentially lossy. I would argue that a `Path` is _conventionally representable_ in UTF-8 just as much as a `ByteStr` is conventionally UTF-8. Maybe you could say that `ByteStr`'s convention is a stronger convention, but that is splitting hairs.

So I think the only reason standing to not impl Display for either type is the `to_string` footgun. So `ByteStr` and `Path` are in the same boat, and so whatever we do for `ByteStr` will create a precedent for `Path`.

The consensus now is to solve this using an inherent `to_string`. I think that is okay, but I want to raise the option one more time that we could have a `lossy_to_string` lint and avoid the extra inherent method. I imagine the existence of competing `to_string` methods will be a source of confusion at times.

**Comment by clarfonthey** [comment]
To be clear, I *believe* that the arguments stated apply to `Path` and `OsStr` as well, just, we're putting that discussion off for now.

**Comment by BurntSushi** [comment]
Yeah personally, I would love to see `Display` on `OsStr` and `Path` if we can apply the same `to_string()` fix there.

> The consensus now is to solve this using an inherent to_string. I think that is okay, but I want to raise the option one more time that we could have a lossy_to_string lint and avoid the extra inherent method. I imagine the existence of competing to_string methods will be a source of confusion at times.

I grant there could be some confusion, but I think `to_string()` is also a great name to use for the conversion itself. It's a great name and it would be a pity to see it go to waste via a lint IMO.

**Comment by glandium** [comment]
> Yeah personally, I would love to see `Display` on `OsStr` and `Path` if we can apply the same `to_string()` fix there.

I don't think I agree. It's already bad enough that build scripts use `println!("cargo:rerun-if-changed={}", path.display())`, and paths are treated casually this way everywhere else, and that would make it even less obvious that something is wrong.

**Comment by programmerjake** [comment]
> It's already bad enough that build scripts use `println!("cargo:rerun-if-changed={}", path.display())`

I expect the source code paths to be sane (UTF-8), supporting non-UTF-8 paths are more for when you have a program that needs to work with any arbitrary path no matter what (e.g. managing files where the name was corrupted on a FAT32 filesystem -- I've had that happen) instead of just sane paths.

So, I think `ByteStr` can support `Display` since that's intended for showing stuff to the user, not for making sure you have the correct bytes in your path to that corrupted file.

**Comment by ErichDonGubler** [comment]
@glandium:

> > Yeah personally, I would love to see `Display` on `OsStr` and `Path` if we can apply the same `to_string()` fix there.
> 
> I don't think I agree. It's already bad enough that build scripts use `println!("cargo:rerun-if-changed={}", path.display())`, and paths are treated casually this way everywhere else, and that would make it even less obvious that something is wrong.

Doesn't this point imply that folks would use `BStr` as a path, despite there already existing the `Path` API? Maybe I'm misunderstanding how you think this would be misapplied?

**Comment by Cheese-Space** [comment]
Would, when ByteStr becomes stable, the type of b"..." be &ByteStr instead of &[u8; N] or would you have to manually convert it into a &ByteStr?

**Comment by ChrisDenton** [comment]
> Would, when ByteStr becomes stable, the type of b"..." be &ByteStr instead of &[u8; N] or would you have to manually convert it into a &ByteStr?

That would be a breaking change so could only be done over an edition. Though I fear that would involve too much churn in user code to be worth it (I could be wrong though).

**Comment by Cheese-Space** [comment]
@ChrisDenton are b"..." strings really used that often in user code? I've only used it once I think but maybe that is because I don't work on no_std projects.

**Comment by BANOnotIT** [comment]
@Cheese-Space 
> are b"..." strings really used that often in user code?

Yes, quick-xml for example [uses it for tag names and attributes](https://docs.rs/quick-xml/latest/quick_xml/name/struct.QName.html). That's needed [to support encodings other than utf8](https://docs.rs/quick-xml/latest/quick_xml/index.html#encoding).

**Comment by camsteffen** [comment]
Maybe `b".."` can resolve to either `&ByteStr` or `&[u8; N]` in the same way that a numeric literal can be different int types. Changing to `&ByteStr` wholesale is a non-starter since you'd lose array length.

**Comment by Cheese-Space** [comment]
@camsteffen that's actually a good idea! we could make it default to &[u8; N] (as that is used currently) but it could resolve to &ByteStr if you wanted to:
``` rust
let a_byte_string: &ByteStr = b"hello world!";
let a_byte_array = b"hello world!"; // type is &[u8; 12]
```

**Comment by glandium** [comment]
> > I don't think I agree. It's already bad enough that build scripts use `println!("cargo:rerun-if-changed={}", path.display())`, and paths are treated casually this way everywhere else, and that would make it even less obvious that something is wrong.
> 
> Doesn't this point imply that folks would use `BStr` as a path, despite there already existing the `Path` API? Maybe I'm misunderstanding how you think this would be misapplied?

I was responding to a comment suggesting to add a `Display` impl on `Path`.

**Comment by DanielEScherzer** [comment]
> Should the Display impl use the Unicode replacement character, or do escaping like the Debug impl?

In #159300 the display impl was moved to a separate feature gate (`bstr_to_string`) so hopefully this doesn't block stabilization of the overall type

**Comment by clarfonthey** [comment]
(FWIW, trait implementations can't be unstable, so, we're still going to need to discuss `Display` on stabilisation. But the methods at least are unstable, and we can comment out the `Display` impl if we need.)

In terms of final changes before stabilisation, I opened #160912 to rename the module to `byte_str` from `bstr`, which we haven't fully discussed but I assumed we wanted. If the type is going to be named `ByteStr` instead of `BStr`, we will want this.

I'll rebase the opaque rewrite on top of that one, but I figured the module/feature rename would be easier to do first.

**Comment by programmerjake** [comment]
> FWIW, trait implementations can't be unstable

that's no longer the case: https://github.com/rust-lang/rust/blob/712070e7e00c742aea222da41bd10f8d0de3dfeb/src/doc/rustc-dev-guide/src/stability.md#unstable_feature_bound

**Comment by clarfonthey** [comment]
That also goes into the entire separate issue that I've been meaning to write up, which is that rustc-dev-guide and std-dev-guide do not have the same information and should probably be merged somehow.
